aten/src/ATen/native/LossMultiMargin.cpp

```diff
@@ -0,0 +1,337 @@
+#include <ATen/ATen.h>
+#include <ATen/Dispatch.h>
+
+namespace at {
+namespace native {
+
+namespace {
+
+template <typename scalar_t>
+inline scalar_t multi_margin_inner_sum_cpu(
+    const scalar_t* input_data,
+    const scalar_t* weight_data,
+    const int p,
+    const scalar_t margin,
+    const int64_t dim,
+    const int64_t target_idx) {
+  const scalar_t input_target = input_data[target_idx];
+  scalar_t sum = 0;
+  for (int64_t d = 0; d < dim; d++) {
+    if (d == target_idx) {
+      continue;
+    }
+
+    const scalar_t z = margin - input_target + input_data[d];
+    if (z > 0) {
+      scalar_t h = (p == 1) ? z : z * z;
+      if (weight_data != nullptr) {
+        h *= weight_data[target_idx];
+      }
+      sum += h;
+    }
+  }
+
+  sum /= dim;
+  return sum;
+}
+
+inline int64_t target_index_checked(
+    const int64_t* target_data,
+    const int64_t index,
+    const int64_t dim) {
+  const int64_t idx = target_data[index];
+  TORCH_CHECK(idx >= 0 && idx < dim, "target out of range");
+  return idx;
+}
+
+template <typename scalar_t>
+static inline void multi_margin_loss_cpu_kernel(
+    Tensor& output,
+    scalar_t* input_data,
+    int64_t* target_data,
+    const int p,
+    scalar_t margin,
+    scalar_t* weight_data,
+    const int64_t nframe,
+    const int64_t dim,
+    const int64_t reduction) {
+  // dim() != 0 check is for 1d input which produces a scalar output (that
+  // cannot be handeld by TensorAccessor)
+  if (reduction == Reduction::None && output.dim() > 0) {
+    auto output_acc = output.accessor<scalar_t, 1>();
+    for (int64_t t = 0; t < nframe; t++) {
+      const auto idx = target_index_checked(target_data, t, dim);
+      auto sum = multi_margin_inner_sum_cpu(
+          input_data, weight_data, p, margin, dim, idx);
+      output_acc[t] = sum;
+      input_data += dim;
+    }
+  } else {
+    scalar_t sum = 0;
```

Comment:
I think we should use accscalar_t instead

Reply:
What is this type? Is it also useful for CPU code?